---
title: "GWAS in R"
author: "Wanqi"
date: 2021-01-20T21:13:54-06:00
output: html_document
---

```{r}
library(rrBLUP)
```

https://link.springer.com/protocol/10.1007%2F978-1-4939-6682-0_14


## data import 
The data provided by the guide is Rdata format, which needs to be imported by load(). 

```{r}
load("data/GWAS_Data.RData")
```

## phenotype

View phenotypic data
```{r}
head(pheno)
```

Check variables structure. GID and ENV needs to be factors
```{r}
str(pheno)
```

Checking normality of the data. In theory residuals needs to be checked 
but in general if data are normal, residuals are normal. 

```{r}

hist(pheno$Yield, col="black",xlab="Yield",ylab="Frequency",
     border="white",breaks=10,main="Yield Histogram") # Data seems pretty normal with some outliers. 
```

```{r}
shapiro.test(pheno$Yield)
```
Shapiro-Wilk test indicates that normality condition is not met. LetÂ´s remove the outliers and check the normality again. 

```{r}
boxplot.yield<- boxplot(pheno$Yield,xlab="Boxplot",ylab="Yield",ylim=c(4000,9000))
```

```{r}
outliers <- boxplot.yield$out
outliers
```
10 outliers, let's remove it

```{r}
pheno <- pheno[-which(pheno$Yield%in%outliers),]
```

```{r}
shapiro.test(pheno$Yield)
```
it is normal now!

let's now remove all missing data
```{r}
pheno <- na.omit(pheno)
```

After filtering pheno 922 observations

## genotype

View genotypic data
```{r}
geno[1:5,1:5]
```

```{r}
map[1:5,1:3]
```


Filtering conditions will depend on the researcher. In this simple, function
we remove individuals with missing data, markers with a certain % of missing,
and heterozygous calls (IF there are a high proportion of heterozygous, it can
indicate a problem with the marker, because oat is an inbreed specie)

```{r}
filter.fun <- function(geno,IM,MM,H){
  #Remove individuals with more than a % missing data
  individual.missing <- apply(geno,1,function(x){
    return(length(which(is.na(x)))/ncol(geno))
  })
  #length(which(individual.missing>0.40)) #will tell you how many 
  #individulas needs to be removed with 20% missing.
  #Remove markers with % missing data
  marker.missing <- apply(geno,2,function(x)
  {return(length(which(is.na(x)))/nrow(geno))
    
  })
  length(which(marker.missing>0.6))
  #Remove markers herteozygous calls more than %. 
  heteroz <- apply(geno,1,function(x){
    return(length(which(x==0))/length(!is.na(x)))
  })
  
  filter1 <- geno[which(individual.missing<IM),which(marker.missing<MM)]
  filter2 <- filter1[,(heteroz<H)]
  return(filter2)
}
geno[1:10,1:10]
```

```{r}
geno.filtered <- filter.fun(geno[,1:3629],0.4,0.60,0.02)
geno.filtered[1:5,1:5]
dim(geno.filtered)
```


## Imputation

rrBLP program will make imputation. For the simplicity , we impute using the mean but EM algorithm can be also used. rrBLUP also allows to remove markers depending on the Minor allele frequency (MAF),in our example we remove those markers with MAF less than 0.05.

*The EM imputation algorithm is based on the multivariate normal distribution and was designed for use with GBS (genotyping-by-sequencing) markers, which tend to be high density but with lots of missing data. Details are given in Poland et al. (2012). The EM algorithm stops at iteration t when the RMS error = n^{-1} \|A_{t} - A_{t-1}\|_2 < tol.*

```{r}
library(rrBLUP)
Imputation <- A.mat(geno.filtered,impute.method="EM",return.imputed=T,min.MAF=0.05)
```
```{r}
K.mat <- Imputation$A ### KINSHIP matrix
geno.gwas <- Imputation$imputed #NEW geno data.
```


```{r}
geno.gwas[1:5,1:5]## view geno
K.mat[1:5,1:5]## view Kinship
```

## CHECKING POPULATION STRUCTURE EFFECTS

Principal components analysis
```{r}
geno.scale <- scale(geno.gwas,center=T,scale=F) # Data needs to be center.
svdgeno <- svd(geno.scale) 
PCA <- geno.scale%*%svdgeno$v #Principal components
PCA[1:5,1:5]
```



## Screeplot to visualize the proportion of variance explained by PCA
pdf("Figure2.pdf")
plot(round((svdgeno$d)^2/sum((svdgeno$d)^2),d=7)[1:10],type="o",main="Screeplot",xlab="PCAs",ylab="% variance")
dev.off()
##Proportion of variance explained by PCA1 and PCA2
PCA1 <- 100*round((svdgeno$d[1])^2/sum((svdgeno$d)^2),d=3); PCA1
PCA2 <- 100*round((svdgeno$d[2])^2/sum((svdgeno$d)^2),d=3); PCA2
### Plotting Principal components.
pdf("Figure3.pdf")
plot(PCA[,1],PCA[,2],xlab=paste("Pcomp:",PCA1,"%",sep=""),ylab=paste("Pcomp:",PCA2,"%",sep=""),pch=20,cex=0.7)
dev.off()
### Plotting depending on clustering. 
Eucl <- dist(geno.gwas) ###Euclinean distance
fit <- hclust(Eucl,method="ward.D2")### Ward criterion makes clusters with same size.
groups2 <- cutree(fit,k=2) ### Selecting two clusters.
table(groups2)# Number of individuals per cluster.
pdf("Figure4.pdf")
plot(PCA[,1],PCA[,2],xlab=paste("Pcomp:",PCA1,"%",sep=""),ylab=paste("Pcomp:",PCA2,"%",sep=""),pch=20,cex=0.7,col=groups2)
legend("bottomright",c("Subpop1: 244","Subpop2: 84"),pch=20,col=(c("black","red")),lty=0,bty="n",cex=1)
dev.off()

